gk/fib.c

```diff
@@ -0,0 +1,1440 @@
+}
+
+/*
+ * XXX What we are doing here is analogous to RCU's synchronize_rcu(),
+ * what suggests that we may be able to profit from RCU. But we are going
+ * to postpone that until we have a better case to bring RCU to Gatekeeper.
+ */
+static void
+synchronize_gk_instances(struct gk_fib *fib, struct gk_config *gk_conf)
+{
+	int i, loop;
+	int num_succ_notified_inst = 0;
+	bool is_succ_notified[gk_conf->num_lcores];
+
+	/* The maximum number of times to try to notify the GK instances. */
+	const int MAX_NUM_NOTIFY_TRY = 3;
+
+	rte_atomic16_init(&fib->num_updated_instances);
+
+	memset(is_succ_notified, false, sizeof(is_succ_notified));
+
+	for (loop = 0; loop < MAX_NUM_NOTIFY_TRY; loop++) {
+		/* Send the FIB entry to the GK mailboxes. */
+		for (i = 0; i < gk_conf->num_lcores; i++) {
+			if (num_succ_notified_inst == gk_conf->num_lcores)
+				break;
+
+			if (!is_succ_notified[i]) {
+				int ret = notify_gk_instance(fib,
+					&gk_conf->instances[i]);
+				if (ret == 0) {
+					is_succ_notified[i] = true;
+					num_succ_notified_inst++;
+				}
+			}
+		}
+	}
+
+	if (num_succ_notified_inst != gk_conf->num_lcores) {
+		RTE_LOG(WARNING, GATEKEEPER,
+			"gk: %s successfully notifies only %d/%d instances\n",
+			__func__, num_succ_notified_inst, gk_conf->num_lcores);
+	}
+
+	/* Wait for all GK instances to synchronize. */
+	while (rte_atomic16_read(&fib->num_updated_instances)
+			< num_succ_notified_inst)
+		rte_pause();
+}
+
+/*
+ * This function is called by del_fib_entry_locked().
+ * Notice that, it doesn't stand on its own, and it's only
+ * a construct to make del_fib_entry_locked() readable.
+ */
+static struct gk_fib *
+check_fib_entry_for_del_locked(
+	struct ip_prefix *ip_prefix, struct ipaddr *ip_prefix_addr,
+	int ip_prefix_len, struct gk_config *gk_conf)
+{
+	int ret = 0;
+	int ip_prefix_present;
+	struct gk_fib *ip_prefix_fib;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+
+	if (ip_prefix_addr->proto == ETHER_TYPE_IPv4) {
+		uint32_t fib_id;
+
+		ip_prefix_present = rte_lpm_is_rule_present(
+			ltbl->lpm, ntohl(ip_prefix_addr->ip.v4.s_addr),
+			ip_prefix_len, &fib_id);
+		if (!ip_prefix_present) {
+			RTE_LOG(WARNING, GATEKEEPER,
+				"gk: delete an non-existent IP prefix (%s)\n",
+				ip_prefix->str);
+			return NULL;
+		}
+
+		ip_prefix_fib = &ltbl->fib_tbl[fib_id];
+	} else if (likely(ip_prefix_addr->proto == ETHER_TYPE_IPv6)) {
+		uint8_t fib_id;
+
+		ip_prefix_present = rte_lpm6_is_rule_present(
+			ltbl->lpm6, ip_prefix_addr->ip.v6.s6_addr,
+			ip_prefix_len, &fib_id);
+		if (!ip_prefix_present) {
+			RTE_LOG(WARNING, GATEKEEPER,
+				"gk: delete an non-existent IP prefix (%s)\n",
+				ip_prefix->str);
+			return NULL;
+		}
+
+		ip_prefix_fib = &ltbl->fib_tbl6[fib_id];
+	} else {
+		RTE_LOG(WARNING, GATEKEEPER,
+			"gk: delete an IP prefix (%s) with unknown IP type %hu\n",
+			ip_prefix->str, ip_prefix->addr.proto);
+		return NULL;
+	}
+
+	if (ip_prefix_addr->proto == ETHER_TYPE_IPv4) {
+		ret = rte_lpm_delete(ltbl->lpm,
```

Comment:
1. Isn't there an error code when the entry doesn't exist? If so, we can drop the call to `rte_lpm_is_rule_present()` above and simplify the code. The same goes for the IPv6 case.

2. Using `lpm_del_route()` instead of `rte_lpm_delete()` may lead to simpler code.